File: src/bnbprob/slpfssp/cpp/job.cpp
#include "job.hpp"

#include <algorithm>
#include <memory>
#include <numeric>

// Helper function to fill latency matrix for a semiline
static Int2D fill_lat(const std::vector<int>& p)
{
    size_t m = p.size();
    Int2D lat;
    lat.resize(m, std::vector<int>(m, 0));
    for (size_t m1 = 0; m1 < m; ++m1)
    {
        for (size_t m2 = 0; m2 < m; ++m2)
        {
            if (m2 + 1 < m1)
            {
                int sum_p = 0;
                for (size_t i = m2 + 1; i < m1; ++i)
                {
                    sum_p += p[i];
                }
                lat[m1][m2] = sum_p;
            }
        }
    }
    return lat;
}

// Initialize r, q, and lat based on p
void Job::initialize(const Int2DPtr& p_)
{
    if (!p_) return;
    size_t L = p_->size();
    r.clear();
    q.clear();
    lat->clear();
    r.reserve(L);
    q.reserve(L);
    lat->reserve(L);
    for (size_t l = 0; l < L; ++l)
    {
        size_t m = (*p_)[l].size();
        r.push_back(std::vector<int>(m, 0));
        q.push_back(std::vector<int>(m, 0));
        lat->push_back(fill_lat((*p_)[l]));
    }
}

// Get total processing time (sum over all semilines and machines)
int Job::get_T() const
{
    if (!p) return 0;
    int total = 0;
    for (const auto& semiline : *p)
    {
        total += std::accumulate(semiline.begin(), semiline.end(), 0);
    }
    return total;
}

// Copy a job (shallow copy of p and lat, deep copy of r and q)
std::shared_ptr<Job> copy_job(const std::shared_ptr<Job>& jobptr)
{
    if (!jobptr) return nullptr;
    return std::make_shared<Job>(
        jobptr->j, jobptr->p, jobptr->r, jobptr->q,
        jobptr->lat, jobptr->m, jobptr->s);
}

// Copy a vector of jobs
std::vector<std::shared_ptr<Job>> copy_jobs(
    const std::vector<std::shared_ptr<Job>>& jobs)
{
    std::vector<std::shared_ptr<Job>> out(jobs.size());
    for (int i = 0; i < jobs.size(); ++i)
    {
        out[i] = std::make_shared<Job>(*jobs[i]);
    }
    return out;  // Return the copied vector
}

File: src/bnbprob/slpfssp/job.py
import copy

import numpy as np
from numpy.typing import NDArray


class Job:
    """Job representation in permutation flow-shop scheduling problem"""

    j: int
    """Job index"""
    p: list[list[int]]
    """Job processing time on each machine of each parallel semiline"""
    r: list[list[int]]
    """
    Job release date on each machine of each parallel semiline
    based on partial solution
    (1 | :math:`r_j` , :math:`q_j` | :math:`sum_{j in J}{w_j C_j}`)
    """
    q: list[list[int]]
    """
    Job delivery time on each machine of each parallel semiline
    based on partial solution
    (1 | :math:`r_j` , :math:`q_j` | :math:`sum_{j in J}{w_j C_j}`)
    """
    lat: list[NDArray[np.int_]]
    """Job latency between two machines of each semiline
    `sum(p_i k < i < m)`"""
    m: list[int]
    """Number of machines on each parallel semiline"""
    L: int
    """Number of parallel semilines"""
    s: int
    """Number of reconciliation machines"""

    def __init__(self, j: int, p: list[list[int]], s: int = 1) -> None:
        self.j = j
        self.p = p
        self.s = s

        L = len(p)
        m = [len(p_i) for p_i in p]
        r = [[0] * m_i for m_i in m]
        q = [[0] * m_i for m_i in m]
        lat = [_fill_lat(p_i) for p_i in p]

        self.r = r
        self.q = q
        self.lat = lat

        self.m = m
        self.L = L

    def __repr__(self) -> str:
        return self._signature

    def __str__(self) -> str:
        return self._signature

    @property
    def _signature(self) -> str:
        return f'Job {self.j}'

    @property
    def T(self) -> int:
        """Total processing time initialization heuristic"""
        return sum(sum(p_i) for p_i in self.p)

    @staticmethod
    def from_job(job: 'Job') -> 'Job':
        return job.copy(deep=False)

    def copy(self, deep: bool = False) -> 'Job':
        if deep:
            return copy.deepcopy(self)
        other = Job.__new__(Job)
        other.j = self.j
        other.p = self.p
        other.r = [r_i.copy() for r_i in self.r]
        other.q = [q_i.copy() for q_i in self.q]
        other.lat = self.lat
        other.m = self.m
        other.L = self.L
        other.s = self.s
        return other


def start_job(j: int, p: list[list[int]], s: int = 1) -> Job:
    return Job(j, p, s)


def _fill_lat(p: list[int]) -> NDArray[np.int_]:
    m = len(p)

    # Resize `lat` to m x m
    lat = np.zeros((m, m), dtype='i')

    # Compute sums
    for m1 in range(m):
        for m2 in range(m):
            if m2 + 1 < m1:  # Ensure range is valid
                sum_p = 0
                for i in range(m2 + 1, m1):
                    sum_p += p[i]
                lat[m1, m2] = sum_p

    return lat
